File: 2025-2/sesion-de-clase-16/SoftProgPersistencia/DAOImpl/Almacen/ProductoDAOImpl.cs
using System;
using System.Data;
using System.Data.Common;
using PUCP.SoftProg.Modelo.Almacen;
using PUCP.SoftProg.Persistencia.DAO.Almacen;

namespace PUCP.SoftProg.Persistencia.DAOImpl.Almacen {
    public class ProductoDAOImpl : BaseDAOImpl<Producto>, IProductoDAO {
        protected override DbCommand ComandoCrear(DbConnection conn, Producto producto) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "insertarProducto";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_nombre", 
                DbType.String, producto.Nombre);
            this.AgregarParametroEntrada(cmd, "@p_unidadMedida", 
                DbType.String, producto.UnidadMedida);
            this.AgregarParametroEntrada(cmd, "@p_precio", 
                DbType.Double, producto.Precio);
            this.AgregarParametroEntrada(cmd, "@p_activo", 
                DbType.Boolean, producto.IsActive);
            this.AgregarParametroSalida(cmd, "@p_id", DbType.Int32);

            return cmd;
        }

        protected override DbCommand ComandoActualizar(DbConnection conn, Producto producto) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "modificarProducto";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_nombre", DbType.String, producto.Nombre);
            this.AgregarParametroEntrada(cmd, "@p_unidadMedida", DbType.String, producto.UnidadMedida);
            this.AgregarParametroEntrada(cmd, "@p_precio", DbType.Double, producto.Precio);
            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, producto.IsActive);
            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, producto.Id);

            return cmd;
        }

        protected override DbCommand ComandoEliminar(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "eliminarProducto";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeer(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarProductoPorId";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeerTodos(DbConnection conn) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "listarProductos";
            cmd.CommandType = CommandType.StoredProcedure;
            return cmd;
        }

        protected override Producto MapearModelo(DbDataReader reader) {
            return new Producto {
                Id = Convert.ToInt32(reader["id"]),
                Nombre = Convert.ToString(reader["nombre"]),
                UnidadMedida = (UnidadMedida)Enum.Parse(
                    typeof(UnidadMedida), 
                    Convert.ToString(reader["unidadMedida"])), 
                Precio = Convert.ToDouble(reader["precio"]), 
                IsActive = Convert.ToBoolean(reader["activo"])
            };
        }
    }
}

File: 2026-1/sesion-de-clase-10/usa-procedimientos-almacenados/SoftProg/Referencia/softprog/daoimpl/almacen/ProductoDAOImpl.java
package pe.edu.pucp.inf30.softprog.daoimpl.almacen;

import java.sql.CallableStatement;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import pe.edu.pucp.inf30.softprog.dao.almacen.ProductoDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.BaseDAO;
import pe.edu.pucp.inf30.softprog.modelo.almacen.Producto;
import pe.edu.pucp.inf30.softprog.modelo.almacen.UnidadMedida;

/**
 *
 * @author eric
 */
public class ProductoDAOImpl extends BaseDAO<Producto> 
        implements ProductoDAO {
    
    @Override
    protected PreparedStatement comandoCrear(Connection conn, Producto modelo) 
            throws SQLException {
        
        String sql = "{call insertarProducto(?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setString("p_nombre", modelo.getNombre());
        cmd.setString("p_unidadMedida", modelo.getUnidadMedida().name());
        cmd.setDouble("p_precio", modelo.getPrecio());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.registerOutParameter("p_id", Types.INTEGER);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoActualizar(Connection conn, 
            Producto modelo) throws SQLException {
        
        String sql = "{call modificarProducto(?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setString("p_nombre", modelo.getNombre());
        cmd.setString("p_unidadMedida", modelo.getUnidadMedida().name());
        cmd.setDouble("p_precio", modelo.getPrecio());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.setInt("p_id", modelo.getId());
        return cmd;
    }

    @Override
    protected PreparedStatement comandoEliminar(Connection conn, Integer id) throws SQLException {
        String sql = "{call eliminarProducto(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_id", id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeer(Connection conn, Integer id) throws SQLException {
        String sql = "{call buscarProductoPorId(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_id", id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeerTodos(Connection conn) throws SQLException {
        String sql = "{call listarProductos()}";
        CallableStatement cmd = conn.prepareCall(sql);
        return cmd;
    }

    @Override
    protected Producto mapearModelo(ResultSet rs) throws SQLException {
        Producto producto = new Producto();
        producto.setId(rs.getInt("id"));
        producto.setNombre(rs.getString("nombre"));
        producto.setUnidadMedida(UnidadMedida.valueOf(rs.getString("unidadMedida")));
        producto.setPrecio(rs.getDouble("precio"));
        producto.setActivo(rs.getBoolean("activo"));
        return producto;
    }
}
